### PR TITLE
Adding managed field for certifcates and some fixes for Key Vault

### PR DESCRIPTION
**Pull request by pomortaz**
1. Adding managed field to key object and secret object for certifciates
2. Fix mandatory parameter certificate issuer and certificate operation
3. Fix jsonWebKey json tag

### CODE REVIEW

**Comment by amarzavery** [comment]
@pomortaz - Please send the pr for swagger spec to the public repo for  kv mgmt plane.

**Comment by pomortaz** [comment]
Thanks @amarzavery. The changes are for key vault data-plane.

**Comment by pomortaz** [comment]
@kunalch and @rohmano please review. :)

File: lib/services/keyVault/lib/keyVault.js [hunk 1]
```diff
@@ -1053,34 +1053,32 @@ KeyVaultClient.prototype.getCertificateIssuers = function (vaultBaseUrl, options
  * 
  * @param {string} issuerName The name of the issuer.
  * 
- * @param {object} issuer The issuer bundle.
+ * @param {string} provider The name of the issuer.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: >The name of the issuer. [](start = 28, length = 23)

This should say - "The issuer provider"

File: lib/services/keyVault/lib/keyVault.js [hunk 2]
```diff
@@ -1109,34 +1107,32 @@ KeyVaultClient.prototype.setCertificateIssuer = function (vaultBaseUrl, issuerNa
  * 
  * @param {string} issuerName The name of the issuer.
  * 
- * @param {object} issuer The issuer bundle.
+ * @param {object} [options] Optional Parameters.
  * 
- * @param {string} [issuer.provider] The name of the issuer.
+ * @param {string} [options.provider] The name of the issuer.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: >The name of the issuer. [](start = 38, length = 23)

Same as above

File: lib/services/keyVault/lib/keyVault.js [hunk 3]
```diff
@@ -1293,7 +1289,7 @@ KeyVaultClient.prototype.deleteCertificateIssuer = function (vaultBaseUrl, issue
  * the issuer of the X509 component of a certificate.
  * 
  * @param {string} [options.certificatePolicy.issuerReference.name] Name of
- * the referenced issuer object.
+ * the referenced issuer object e.g. 'Self', 'Unknown'.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: >the referenced issuer object e.g. 'Self', 'Unknown'. [](start = 3, length = 52)

Either name of the referenced issuer object or one of the reserved name of 'Self' or 'Unknown'

File: lib/services/keyVault/lib/keyVault.js [hunk 4]
```diff
@@ -1414,7 +1410,7 @@ KeyVaultClient.prototype.createCertificate = function (vaultBaseUrl, certificate
...
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: Same as above.

File: lib/services/keyVault/lib/keyVault.js [hunk 5]
```diff
@@ -1587,7 +1583,7 @@ KeyVaultClient.prototype.getCertificatePolicy = function (vaultBaseUrl, certific
  * of the X509 component of a certificate.
  * 
  * @param {string} [certificatePolicy.issuerReference.name] Name of the
- * referenced issuer object.
+ * referenced issuer object e.g. 'Self', 'Unknown'.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: >referenced issuer object e.g. 'Self', 'Unknown'. [](start = 3, length = 48)

Same as above.

File: lib/services/keyVault/lib/keyVault.js [hunk 6]
```diff
@@ -1621,13 +1617,90 @@ KeyVaultClient.prototype.updateCertificatePolicy = function (vaultBaseUrl, certi
...
+ * key backing a certificate.
+ * 
+ * @param {boolean} [options.certificatePolicy.keyProperties.exportable]
+ * Indicates if the private key can be exported.
+ * 
+ * @param {string} [options.certificatePolicy.keyProperties.keyType] The key
+ * type.
+ * 
+ * @param {number} [options.certificatePolicy.keyProperties.keySize] The key
+ * size in bytes. e.g. 1024 or 2048.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: >1024 or [](start = 23, length = 7)

only 2048 is allowed for certificates

File: lib/services/keyVault/lib/keyVault.js [hunk 7]
```diff
@@ -1621,13 +1617,90 @@ KeyVaultClient.prototype.updateCertificatePolicy = function (vaultBaseUrl, certi
...
+ * 
+ * @param {object} [options.certificatePolicy.x509CertificateProperties]
+ * Properties of the X509 component of a certificate.
+ * 
+ * @param {string}
+ * [options.certificatePolicy.x509CertificateProperties.subject] The subject
+ * name. Should be a valid X509 Distinguished Name.
+ * 
+ * @param {array} [options.certificatePolicy.x509CertificateProperties.ekus]
+ * The enhaunced key usage.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: >enhaunced [](start = 7, length = 9)

spell check.

Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: I think we are missing one field of key usages.

File: lib/services/keyVault/lib/keyVault.js [hunk 8]
```diff
@@ -1621,13 +1617,90 @@ KeyVaultClient.prototype.updateCertificatePolicy = function (vaultBaseUrl, certi
...
+ * [options.certificatePolicy.x509CertificateProperties.subjectAlternativeNames.upns]
+ * User principal names.
+ * 
+ * @param {array}
+ * [options.certificatePolicy.x509CertificateProperties.keyUsage] List of key
+ * usages.
+ * 
+ * @param {number}
+ * [options.certificatePolicy.x509CertificateProperties.validityInMonths] The
+ * subject alternate names.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: Wrong comment.

File: lib/services/keyVault/lib/keyVault.js [hunk 9]
```diff
@@ -1621,13 +1617,90 @@ KeyVaultClient.prototype.updateCertificatePolicy = function (vaultBaseUrl, certi
...
+ * 
+ * @param {array} [options.certificatePolicy.lifetimeActions] Actions that
+ * will be performed by Key Vault over the lifetime of a certificate.
+ * 
+ * @param {object} [options.certificatePolicy.issuerReference] Reference to
+ * the issuer of the X509 component of a certificate.
+ * 
+ * @param {string} [options.certificatePolicy.issuerReference.name] Name of
+ * the referenced issuer object e.g. 'Self', 'Unknown'.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVault.js]: same as one of the previous comments on this.

File: lib/services/keyVault/lib/keyVaultClient.d.ts [hunk 10]
```diff
@@ -937,7 +935,7 @@ declare class KeyVaultClient {
          * the issuer of the X509 component of a certificate.
          * 
          * @param {string} [options.certificatePolicy.issuerReference.name] Name of
-         * the referenced issuer object.
+         * the referenced issuer object e.g. 'Self', 'Unknown'.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVaultClient.d.ts]: >the referenced issuer object e.g. 'Self', 'Unknown'. [](start = 11, length = 52)

Same comment as previous comment on this topic.

File: lib/services/keyVault/lib/keyVaultClient.d.ts [hunk 11]
```diff
@@ -1223,6 +1221,76 @@ declare class KeyVaultClient {
...
+         * key backing a certificate.
+         * 
+         * @param {boolean} [options.certificatePolicy.keyProperties.exportable]
+         * Indicates if the private key can be exported.
+         * 
+         * @param {string} [options.certificatePolicy.keyProperties.keyType] The key
+         * type.
+         * 
+         * @param {number} [options.certificatePolicy.keyProperties.keySize] The key
+         * size in bytes. e.g. 1024 or 2048.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVaultClient.d.ts]: >1024 [](start = 31, length = 4)

1024 is not supported.

File: lib/services/keyVault/lib/keyVaultClient.d.ts [hunk 12]
```diff
@@ -1223,6 +1221,76 @@ declare class KeyVaultClient {
...
+         * 
+         * @param {array}
+         * [options.certificatePolicy.x509CertificateProperties.keyUsage] List of key
+         * usages.
+         * 
+         * @param {number}
+         * [options.certificatePolicy.x509CertificateProperties.validityInMonths] The
+         * subject alternate names.
+         * 
+         * @param {array} [options.certificatePolicy.lifetimeActions] Actions that
```
Comment (kunalcha) [inline lib/services/keyVault/lib/keyVaultClient.d.ts]: >lifetimeActions [](start = 53, length = 15)

this is an array of complex objects. Is this defined anywhere? Should it be?

File: lib/services/keyVault/lib/models/certificateBundle.js [hunk 13]
```diff
@@ -87,7 +87,7 @@ var models = require('./index');
  * X509 component of a certificate.
  * 
  * @member {string} [policy.issuerReference.name] Name of the referenced
- * issuer object.
+ * issuer object e.g. 'Self', 'Unknown'.
```
Comment (kunalcha) [inline lib/services/keyVault/lib/models/certificateBundle.js]: >issuer object e.g. 'Self', 'Unknown'. [](start = 3, length = 37)

Either name of the referenced issuer object or one of the reserved name of 'Self' or 'Unknown'

File: lib/services/keyVault/lib/models/jsonWebKey.js [hunk 14]
```diff
@@ -149,7 +150,7 @@ JsonWebKey.prototype.mapper = function () {
         },
         k: {
           required: false,
-          serializedName: 'K',
+          serializedName: 'k',
```
Comment (kunalcha) [inline lib/services/keyVault/lib/models/jsonWebKey.js]: >'k' [](start = 26, length = 3)

Is this Intentional?

Comment (kunalcha) [inline lib/services/keyVault/lib/models/jsonWebKey.js]: Seems ok to me, just checking.

---
In reply to: [76497637](https://github.com/Azure/azure-sdk-for-node/pull/1858#discussion_r76497637) [](ancestors = 76497637)

Comment (kunalcha) [inline lib/services/keyVault/lib/models/jsonWebKey.js]: Can this cause serialization breaking change for folks who have serialized the key?

---
In reply to: [76497661](https://github.com/Azure/azure-sdk-for-node/pull/1858#discussion_r76497661) [](ancestors = 76497661,76497637)

File: test/services/keyVault/keyVault-certificate-tests.js [hunk 15]
```diff
@@ -426,7 +426,7 @@ describe('Key Vault certificates', function () {
       };
 
       function createCertificateIssuer(next) {
-        client.setCertificateIssuer(vaultUri, ISSUER_NAME, issuerBundle, function (err, responseIssuerBundle) {
+        client.setCertificateIssuer(vaultUri, ISSUER_NAME, 'test', issuerBundle, function (err, responseIssuerBundle) {
```
Comment (kunalcha) [inline test/services/keyVault/keyVault-certificate-tests.js]: >'test', [](start = 59, length = 7)

Is this a provider name of the test issuer provider?

Comment (pomortaz) [inline test/services/keyVault/keyVault-certificate-tests.js]: Yes

File: lib/services/keyVault/lib/models/jsonWebKey.js (hunk 14 above)
Comment (pomortaz) [inline lib/services/keyVault/lib/models/jsonWebKey.js]: It is a bug fix. It wouldn't get deserialize without this change for nodeJs.

---
In reply to: [76497731](https://github.com/Azure/azure-sdk-for-node/pull/1858#discussion_r76497731) [](ancestors = 76497731,76497661,76497637)

File: lib/services/keyVault/lib/keyVaultClient.d.ts (hunk 12 above)
Comment (pomortaz) [inline lib/services/keyVault/lib/keyVaultClient.d.ts]: Yes in the models: [https://github.com/Azure/azure-sdk-for-node/blob/master/lib/services/keyVault/lib/models/lifetimeAction.js](https://github.com/Azure/azure-sdk-for-node/blob/master/lib/services/keyVault/lib/models/lifetimeAction.js)


---
In reply to: [76497264](https://github.com/Azure/azure-sdk-for-node/pull/1858#discussion_r76497264) [](ancestors = 76497264)

**Comment by kunalcha** [comment]
:shipit:

**Comment by amarzavery** [comment]
@pomortaz - Have you updated the package.json?

**Pull request merged**